usdUtils: add USDZ Package Encapsulation Validator

beersandrew opened the pull request:
### Description of Change(s)
- Adding all necessary validator files to support a validator in usdUtils
- Added Initial code for USDZ package encapsulation validation

### Fixes Issue(s)
- https://github.com/orgs/PixarAnimationStudios/projects/3/views/1?pane=issue&itemId=65424899

<!--
Please follow the Contributing and Building guidelines to run tests against your
change. Place an X in the box if tests are run and are all tests passing.
-->
- [X] I have verified that all unit tests pass with the proposed changes
<!-- 
Place an X in the box if you have submitted a signed Contributor License Agreement.
A signed CLA must be received before pull requests can be merged.
For instructions, see: http://openusd.org/release/contributing_to_usd.html
-->
- [X] I have submitted a signed Contributor License Agreement

beersandrew committed:
usd: Initial files to create usdz validation

- added template code for usdz validation
- referenced existing code in usdUtils to do the work of checking if all files are encapsulated

NOTE: this branch fails to link currently

beersandrew committed:
remove versions in types

beersandrew committed:
feat: move usdzValidator to usdUtils

- moved all files from usd -> usdUtils
- added initial test for packageEncapsulation

beersandrew committed:
fix: remove usdzValidators from usd

- clean up validatorTokens from usd

beersandrew commented on pxr/usd/usdUtils/testenv/testUsdUtilsValidators.cpp:
```diff
@@ -0,0 +1,94 @@
+    const std::string testAssetDirectory = ArchGetCwd() + "/ctest/testUsdUtilsPackageEncapsulation/";
+
+    // Load the pre-created usdz stage with fully qualified paths to both a reference and an asset
+    const UsdStageRefPtr& stage = UsdStage::Open(testAssetDirectory + "fail.usdz");
```
@tallytalwar I'm looking for feedback on this section. I'd like to do this in a way that does not depend on so many test files. I tried using [UsdUtilsCreateNewUsdzPackage](https://github.com/PixarAnimationStudios/OpenUSD/blob/dev/pxr/usd/usdUtils/usdzPackage.cpp#L81) however when using that on my `usda` file with fully qualified paths, the function corrects the paths internally so I don't get the failure condition I want. 

My next idea was to manually set the paths in code:

1. Create `subLayer.usda` (which is empty)
2. Create `image.jpg` (or some other asset file that's mostly empty)
3. Create `root.usda` that references the fully qualified paths of `subLayer.usda` & `image.jpg`
4. "Manually" turn `root.usda` into `test.usdz` (not sure what this function is, but not the above function since that corrects my paths, `usdzip` seems to work for this so maybe I can do the equivalent of that)
5. Verify errors
6. Update reference and asset paths in root.usda to be relative paths
7. "Manually" turn `root.usda` into `test.usdz` again (I guess here I could run `UsdUtilsCreateNewUsdzPackage` however to me it would be more clear to create the usdz the same way for someone reading the function
8. Verify the errors are gone
9. Delete `subLayer.usda`, `image.jpg`, 'root.usda', and `test.usdz`

This path requires no tests assets which seems better to me but I wanted to get your thoughts before going forward incase you like the committed solution better or had some other idea

tallytalwar replied:
I am not sure I am groking fully the aim here? Lets talk on slack what you want to achieve here.

beersandrew committed:
refactor: clean up whitespace

jesschimein commented:
Filed as internal issue #USD-9933

jesschimein commented:
/AzurePipelines run

tallytalwar commented:
Will get to reviewing this soon @beersandrew. It will be good to also get @dgovil 's eyes on any usdz related validator, so including him. Thanks both.

tallytalwar commented on pxr/usd/usd/plugInfo.json:
```diff
@@ -148,7 +148,7 @@
                 "Validators": {
                     "CompositionErrorTest": {
                         "doc": "Validator aims at providing all composition errors, which were generated while composing the stage."
-                    }, 
```
We have some code internally which checks if any generated code differs.. unfortunately this change will make that fail, can you revert this file?

tallytalwar commented on pxr/usd/usdUtils/testenv/testUsdUtilsValidators.cpp:
```diff
@@ -0,0 +1,94 @@
+TestUsdUsdzValidators()
+{
+    // This should be updated with every new validator added with
+    // UsdUsdzValidators keyword.
```
I think you meant UsdUtilsValidators here in the comment?

tallytalwar commented on pxr/usd/usdUtils/testenv/testUsdUtilsValidators.cpp:
```diff
@@ -0,0 +1,94 @@
+
+    TF_AXIOM(validator);
+
+    const std::string testAssetDirectory = ArchGetCwd() + "/ctest/testUsdUtilsPackageEncapsulation/";
```
So I don't think you need to do this. Refer this example: `testUsdTimeValueAuthoring`,
cmake install dir: https://github.com/PixarAnimationStudios/OpenUSD/blob/release/pxr/usd/usd/CMakeLists.txt#L681-L684
test usage of the installed usda files:
https://github.com/PixarAnimationStudios/OpenUSD/blob/release/pxr/usd/usd/testenv/testUsdTimeValueAuthoring.cpp#L81
And the files are directly used as in the test, that is because thats where ctest will place the test binary and the installed file. 

And doing this way of hard coding ctest path will make this fail internally, as we do not have cmake based build system internally.

tallytalwar replied:
In this code you should be able to use "testUsdUtilsPackageEncapsulation/fail.usdz" directly with UsdStage::Open.

tallytalwar commented on pxr/usd/usdUtils/testenv/testUsdUtilsValidators.cpp:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    // Load the pre-created usdz stage with relative paths to both a reference and an asset.
+    const UsdStageRefPtr &passStage = UsdStage::Open(testAssetDirectory + "pass.usdz");
```
Similar to fail.usdz, you should be able to provide the path: "testUsdUtilsPackageEncapsulation/pass.usdz" here.

tallytalwar commented on pxr/usd/usdUtils/testenv/testUsdUtilsValidators.cpp:
```diff
@@ -0,0 +1,94 @@
+    TestUsdUsdzValidators();
+    TestPackageEncapsulationValidator();
+
+    std::cout << "OK\n";
```
I have been removing these cout statements internally.

tallytalwar commented on pxr/usd/usdUtils/validators.cpp:
```diff
@@ -0,0 +1,88 @@
+
+PXR_NAMESPACE_OPEN_SCOPE
+
+    static
```
fix indentation.

tallytalwar commented on pxr/usd/usdUtils/validators.cpp:
```diff
@@ -0,0 +1,88 @@
+
+        if (packagePath.length() > 0) {
+            for (const SdfLayerRefPtr& subLayer : layers) {
+                const std::string& realPath = subLayer->GetRealPath();
```
I was comparing this implementation with the ComplianceChecker implementation, and I saw a note about skipping session layers, which won't have a realPath, seems thats missing here.
https://github.com/PixarAnimationStudios/OpenUSD/blob/release/pxr/usd/usdUtils/complianceChecker.py#L753-L755

tallytalwar commented on pxr/usd/usdUtils/validators.cpp:
```diff
@@ -0,0 +1,88 @@
+                                        ArSplitPackageRelativePathOuter(realPath).first :
+                                        realPath;
+
+        if (packagePath.length() > 0) {
```
`if (!packagePath.empty())`

tallytalwar commented on pxr/usd/usdUtils/validators.cpp:
```diff
@@ -0,0 +1,88 @@
+                                    UsdValidationErrorSite(rootLayer,
+                                                           subLayer->GetDefaultPrimAsPath())
+                            },
+                            TfStringPrintf(("Found loaded layer '%s' that "
```
I know the error message is coming from the current compliance check code. But why "loaded" layer? Maybe @dgovil or @spiffmon can clarify?

beersandrew replied:
I'm second guessing my decision to remove 'loaded' in the error message, also interested about it's meaning in this context.

tallytalwar reviewed (commented):
Thanks @beersandrew. Left a few notes and feel free to ping on slack with the test question you had.

beersandrew committed:
fix: various updates from PR comments

- fixed indentation
- fixed misnamed validators in comments
- use .empty() instead of size() > 0
- fix usd/plugInfo.json
- remove test print statement

tallytalwar commented on pxr/usd/usdUtils/validators.cpp:
```diff
@@ -0,0 +1,88 @@
+        return errors;
+    }
+
+    std::vector<TfRefPtr<SdfLayer>> layers;
```
You should be able to use SdfLayerRefPtrVector, defined here: https://github.com/PixarAnimationStudios/OpenUSD/blob/release/pxr/usd/sdf/declareHandles.h#L335

beersandrew committed:
fix: various PR comment related fixes

- instead of using a FQP for the failure condition, changed to using a relative path that was not actually included, making verifying error messages easier
- used better typing where possible
- remove the word 'loaded' from error message
- fixed indenting in CMakeLists.txt

beersandrew committed:
fix: do not include in-memory layers

- in-memory layers will not have a real path, exclude them in this validator

beersandrew commented:
Thanks for the notes @tallytalwar I think I've addressed all of them.

tallytalwar commented on pxr/usd/usdUtils/testenv/testUsdUtilsValidators.cpp:
```diff
@@ -47,24 +47,25 @@ TestPackageEncapsulationValidator()
-
-    // Load the pre-created usdz stage with fully qualified paths to both a reference and an asset
-    const UsdStageRefPtr& stage = UsdStage::Open(testAssetDirectory + "fail.usdz");
+    // Load the pre-created usdz stage with paths to a layer and asset that are not included in the package, but exist
```
80 column width limit please.

jesschimein commented:
/AzurePipelines run

beersandrew committed:
refactor: fix long lines of code

- adjust all lines of code such that they are no wider than column 80

tallytalwar commented on pxr/usd/usdUtils/validators.cpp:
```diff
@@ -0,0 +1,90 @@
+            const std::string& realPath = subLayer->GetRealPath();
+
+            // Make sure the realPath exists, as the validator should not consider in-memory layers
+            if (!realPath.empty() && !TfStringStartsWith(realPath, packagePath)) {
```
nit:

I would have done:
```cpp
if (realPath.empty()) {
    continue;
}
if (!TfStringStartsWith(realPath, packagePath)) {
....
}
```

beersandrew replied:
thanks! I agree that's clearer, updated

tallytalwar reviewed (approved):
LGTM @beersandrew Thanks

beersandrew committed:
refactor: make skipping session layers clearer

tallytalwar commented on pxr/usd/usdUtils/testenv/testUsdUtilsValidators.cpp:
```diff
@@ -0,0 +1,102 @@
+    const std::string& rootDirectory = realRootPath
+            .substr(0, realRootPath.find_last_of("/") + 1);
+
+    const std::string& errorLayer = rootDirectory + "excludedDirectory"
```
In the validator code you have the error string use identifier, which is correct. But here you are using realPath, which is platform specific. So for windows this will fail as the expected error will be having a "\\" and ValidationError string will be using the identifier path which always use the PathDelimiter ("/").

tallytalwar replied:
And as mentioned ^, this does fail on windows in internal test runs.

beersandrew replied:
I've updated this to use std::filesystem to build the paths properly, let me know if there's a better way. It seems when I just try to use the identifier of the root stage, it uses the relative path "pass.usdz" or "fail.usdz" but the error message from the code ends up including the full actual path in the error message, using the realPath of the root usdz file is the best way I can think of to get that root, however since it's known the file will be in the current working directory, maybe it makes sense to use something like `ArchGetCwd()` but I still think i need `std::filesystem` in order to append `excludedDirectory` and get the correct "/" or "\" afterwards.

let me know what you think

tallytalwar replied:
Hi @beersandrew 

Do you think https://github.com/PixarAnimationStudios/OpenUSD/blob/dev/pxr/base/tf/pathUtils.h#L80-L88 might be helpful?

beersandrew replied:
appreciate your patience on this one @tallytalwar, I've updated the pathing to use Tf functions where it seemed appropriate, let me know what you think

tallytalwar commented on pxr/usd/usdUtils/CMakeLists.txt:
```diff
@@ -290,6 +309,11 @@ pxr_install_test_dir(
 )
 
+pxr_install_test_dir(
+    SRC testenv/testUsdUtilsPackageEncapsulation
```
nit: We generally keep the test asset directory same name as the test. So "testUsdUtilsValidators", instead of "testUtilsPackageEncapsulation".

tallytalwar commented on pxr/usd/usdUtils/testenv/testUsdUtilsValidators.cpp:
```diff
@@ -0,0 +1,102 @@
+    // Load the pre-created usdz stage with paths to a layer and asset
+    // that are not included in the package, but exist
+    const UsdStageRefPtr& stage = UsdStage::Open(
+            "testUsdUtilsPackageEncapsulation/fail.usdz");
```
The assets are being "installed" in the test dir, so the file here shouldn't include the test dir. Just "fail.usdz" should pick the asset. This was the reason for test failures on all platforms internally.

tallytalwar commented on pxr/usd/usdUtils/testenv/testUsdUtilsValidators.cpp:
```diff
@@ -0,0 +1,102 @@
+    // Load the pre-created usdz stage with relative paths to both a reference
+    // and an asset that are included in the package.
+    const UsdStageRefPtr& passStage = UsdStage::Open(
+            "testUsdUtilsPackageEncapsulation/pass.usdz");
```
As mentioned above for fail.usda, this need to be updated to just "pass.usdz". 

> The assets are being "installed" in the test dir, so the file here shouldn't include the test dir. Just "fail.usdz" should pick the asset. This is the reason for test failures.

tallytalwar commented on pxr/usd/usdUtils/validators.cpp:
```diff
@@ -0,0 +1,95 @@
+                                UsdValidationErrorSite(rootLayer,
+                                                       subLayer->GetDefaultPrimAsPath())
+                        },
+                        TfStringPrintf(("Found layer '%s' that "
```
Had a chat with Spiff internally. "Loaded" is not correct, but maybe it should be "used"/"referenced". Lets do "Found referenced layer...."

tallytalwar commented on pxr/usd/usdUtils/validators.cpp:
```diff
@@ -0,0 +1,95 @@
+                                    realPath;
+
+    if (!packagePath.empty()) {
+        for (const SdfLayerRefPtr& subLayer : layers) {
```
Spiff also correctly pointed and I agree that the use of variable name is incorrect, "subLayer" doesn't signify the concept of used/referenced "dependencies" of the usdz package. Lets rename this to be "referencedLayer".

tallytalwar reviewed (commented):
I think 1 last set of notes @beersandrew. Can you please address these?

Thanks

beersandrew committed:
fix: various PR comment fixes

- changed test asset folder name to match the name of the cpp file
- updated error message to use "referenced" to describe the layers
- updated variable naming
- updated the way to verify the error messages in the test, utilizing std::filesystem

beersandrew committed:
fix: use Tf where possible for pathing

- avoid GetRealPath and use TfGetAbsPath instead
- concat with TfStringCatPaths, instead of with std::filesystem functions

jesschimein commented:
/AzurePipelines run

tallytalwar commented on pxr/usd/usdUtils/testenv/testUsdUtilsValidators.cpp:
```diff
@@ -55,24 +55,21 @@ TestPackageEncapsulationValidator()
+    const std::string& errorLayer = TfStringCatPaths(
+            usdzRootDirectory, "excludedDirectory/layer.usda");
+    const std::string& errorAsset = TfStringCatPaths(
+            usdzRootDirectory, "excludedDirectory/image.jpg");
```
For Identifier its returning the const reference of the saved layer's identifier, but for all other calls I won't do a reference. Just a `const string` should be fine.

The use of reference for TfGetPathName, TfStringCatPaths is actually incorrect, as it will to get a reference of a temporary string.

beersandrew committed:
fix: don't use references on temp strings

tallytalwar commented on pxr/usd/usdUtils/testenv/testUsdUtilsValidators.cpp:
```diff
@@ -0,0 +1,100 @@
+
+    std::vector<std::string> expectedErrorMessages = {
+            TfStringPrintf(("Found referenced layer '%s' that does not "
+                            "belong to the package '%s'."),
```
Now the layer's path is good, but notice that you are using real path in the validator code for packagePath in the error string, I think which is correct so this tests should also not normalize the packagePath (rootLayer's realPath).

beersandrew replied:
thanks, I've updated this to be the same as the validator code. I wasn't sure if I should bring over: 

`const std::string& packagePath = ArIsPackageRelativePath(rootLayer->GetIdentifier()) ?
                                    ArSplitPackageRelativePathOuter(realPath).first :
                                    realPath;`
                                    
 but since it will always be the realPath for these test files, I took it out. Let me know if I should add it back in or not.

jesschimein commented:
/AzurePipelines run

beersandrew committed:
fix: update error message

- fixed the error message for the packagePath to be the same as the actual code, being based off of the realPath

beersandrew committed:
fix: remove unnecessary check
